scene/gui/tree.cpp

```diff
@@ -4191,12 +4191,9 @@ void Tree::item_selected(int p_column, TreeItem *p_item) {
 }
 
 void Tree::item_deselected(int p_column, TreeItem *p_item) {
-	if (selected_item == p_item) {
+	if (selected_item == p_item && selected_col == p_column) {
 		selected_item = nullptr;
-
-		if (selected_col == p_column) {
-			selected_col = -1;
-		}
+		selected_col = -1;
 	}
```

Comment:
Everything else looks good, and I've tested it and it seems to fix the issue.

This is the only thing that worries me a little, despite it making more sense to me this way. However, it changes how the method works, and I'm not certain there aren't other parts of the code that rely on it working this way.

Is there any situation where we want `Tree::selected_item` to be null, but `Tree::selected_col` to be not null? I can't really think of one, so your change makes sense to me. @KoBeWi was the last person to look at this code - any insights?

Reply:
I also couldn't find a dialog that made use of `Tree::SELECT_SINGLE`, where individual cells of a `TreeItem` might be selected. That would help us test whether selecting/deselecting individual cells causes undesired behavior.

Reply:
I should have created a separate PR for this change. While fixing, I noticed this minor problem, so I also corrected it as it also applies to deselecting elements. In the case of a tree with several columns, if we had a selected cell with several columns, then before this change, deselecting a different cell than the one the cursor is pointing to would also unselect the cursor in the same row, which should not happen the same way as it is with selecting other fields than the cursor is pointing to.
I just tested it by changing the Select type for filesystem_dock and everything works correctly, but it has only one column. When I wanted to test it by creating Tree node with 2 columns in the editor I noticed that the function deselect_all() is not available in GDScript, which I think could also be added.

Should I leave this PR as is or separate them? I just started to contribute so I'm still learning the process of it. In future I will try to separate each change.

Reply:
For `Tree::SELECT_SINGLE` this change should not matter at all as the cursor is always the same as the selected cell so unselecting it will always clear both `Tree::selected_item` and `Tree::selected_col` before and after the change.

Reply:
Yes, I'm having the same issue trying to find a tree with multiple columns to test this out with. I looked through the source code but I think there's only one or two examples of that in the codebase, but with the time I had couldn't really find them & test.

I think it's worth testing multi-column trees. Exposing a few more methods to GDScript doesn't sound like a bad idea to me. I'll let someone else decide whether that ought to go in a new PR, but you can always temporarily do that to test.

From the time I spent looking at this, Tree just seems to be in need of a nice review, documentation & refactor :) Lots of code commended away without explanation.

Reply:
Simple deselecting by mouse looks ok.
Before:
(You can see the problem; deselecting column 1 will remove cursor from column 0)

https://user-images.githubusercontent.com/23726629/201540070-cd1ea3a0-1bf9-48da-b261-65be7bb1973d.mp4

After:

https://user-images.githubusercontent.com/23726629/201539864-bcf3732d-3df4-4c34-bc26-4c697233be00.mp4

Reply:
Wonderful! I'd be happy to see this in :)